### DISCUSSION TITLE
Using Protocol with **kwargs but more specific implementations

### DISCUSSION THREAD

**Discussion by knyghty** [microsoft/pyright#3818] [Q&A] [answered]
I have transformed my code into this MRE that demonstrates the problem:

```python
from typing import Protocol

class Handler(Protocol):
    def __call__(self, message: str, **kwargs: object) -> str:
        ...


def run_handler(handler: Handler, **kwargs: object) -> str:
    return handler(**kwargs)


def my_handler(message: str, extra_message: str) -> str:
    return f"{message} {extra_message}"


print(run_handler(my_handler, message="Hello", extra_message="World"))
```

In reality there are various handlers that take 0...n keyword arguments of varying types.

This gives two errors:
```
  /.../test.py:9:22 - error: Argument of type "object" cannot be assigned to parameter "message" of type "str" in function "__call__"
    "object" is incompatible with "str" (reportGeneralTypeIssues)
  /.../test.py:16:19 - error: Argument of type "(message: str, extra_message: str) -> str" cannot be assigned to parameter "handler" of type "Handler" in function "run_handler"
    Type "(message: str, extra_message: str) -> str" cannot be assigned to type "(message: str, *, **kwargs: object) -> str"
      Parameter 3: type "object" cannot be assigned to type "str"
        "object" is incompatible with "str"
      Parameter "**kwargs" has no corresponding parameter (reportGeneralTypeIssues)
2 errors, 0 warnings, 0 informations
```

The first error I can fix by using `**kwargs: Any` though I'd prefer not to. I'm not sure if there's anything I can do to fix this? The second one is the one I'm really struggling with. Mypy handles this just fine but I still think I might be doing something wrong rather than it being a pyright problem, so I thought I'd ask here first.

**Comment by erictraut** [comment, marked as answer]
Pyright is correct here. The `run_handler` indicates that it expects to receive a function that accepts an arbitrary number of keyword arguments that are of type object, but you're passing it a function that is much more limited. If `run_handler` were to pass it an arbitrary number of keyword arguments, the program would crash.

One potential solution here is to use PEP 612's `ParamSpec` feature:

```python
from typing import ParamSpec, Protocol

P = ParamSpec("P")

class Handler(Protocol[P]):
    def __call__(self, message: str, *args: P.args, **kwargs: P.kwargs) -> str:
        ...

def run_handler(handler: Handler[P], message: str, *args: P.args, **kwargs: P.kwargs) -> str:
    return handler(message, *args, **kwargs)
```
